src/quo2/foundations/colors.cljs

```diff
@@ -260,7 +257,34 @@
   [s]
   (and (string? s) (string/starts-with? s "#")))
 
-(def custom-color
+(def resolve-color
+  "(resolve-color color theme opacity)
+   color   hex string or keyword (resolves from custom, network and semantic colors)
+   theme  :light/:dark
+   opacity 0-100 (optional) - if set theme is ignored and goes to 50 suffix internally"
+  (memoize
+   (fn
+     ([color]
+      (resolve-color color nil nil))
+     ([color theme-or-opacity]
+      (if (keyword? theme-or-opacity)
```

Comment:
I believe this function is more complicated than it needs to be. The following diff shows a simplified version that passes the unit tests.

I did 3 things:

1. Normalize the `colors-map` var where all colors are maps that can vary based on the 50/60 opacity. Once normalized, we don't need special conditions to check if the color is a string hex or not. There's no perf hit here, because the function to resolve the color is memoized anyway.
2. Made the 2-arity version simpler by not checking if the second arg is a keyword or not. I think it's just fine to pass nil as the second argument when calling `(resolve-color some-color nil 50)`. I prefer this way because there's no surprise. The dev doesn't need to remember that the second arg may be an opacity or theme. Mixing types can be confusing. I don't feel strong about this point, but I tend to prefer explicitness.
3. Extracted the non-memoized function. It's common to prefix with `do-`, but others might prefer other names.

```diff
@@ -224,14 +224,14 @@
 
 ;;;; Networks
 (def ^:private networks
-  {:ethereum "#758EEB"
-   :optimism "#E76E6E"
-   :arbitrum "#6BD5F0"
-   :zkSync   "#9FA0FE"
-   :hermez   "#EB8462"
-   :xDai     "#3FC0BD"
-   :polygon  "#AD71F3"
-   :unknown  "#EEF2F5"})
+  {:ethereum {50 "#758EEB" 60 "#758EEB"}
+   :optimism {50 "#E76E6E" 60 "#E76E6E"}
+   :arbitrum {50 "#6BD5F0" 60 "#6BD5F0"}
+   :zkSync   {50 "#9FA0FE" 60 "#9FA0FE"}
+   :hermez   {50 "#EB8462" 60 "#EB8462"}
+   :xDai     {50 "#3FC0BD" 60 "#3FC0BD"}
+   :polygon  {50 "#AD71F3" 60 "#AD71F3"}
+   :unknown  {50 "#EEF2F5" 60 "#EEF2F5"}})
 
 (def ^:private colors-map
   (merge {:primary {50 primary-50
@@ -257,31 +257,26 @@
   [s]
   (and (string? s) (string/starts-with? s "#")))
 
+(defn- do-resolve-color
+  ([color]
+   (do-resolve-color color nil nil))
+  ([color theme]
+   (do-resolve-color color theme nil))
+  ([color theme opacity]
+   (let [suffix         (cond opacity         50
+                              (= :dark theme) 60
+                              :else           50)
+         resolved-color (get-in colors-map [color suffix] color)]
+     (if opacity
+       (alpha resolved-color (/ opacity 100))
+       resolved-color))))
+
 (def resolve-color
   "(resolve-color color theme opacity)
    color   hex string or keyword (resolves from custom, network and semantic colors)
    theme  :light/:dark
    opacity 0-100 (optional) - if set theme is ignored and goes to 50 suffix internally"
-  (memoize
-   (fn
-     ([color]
-      (resolve-color color nil nil))
-     ([color theme-or-opacity]
-      (if (keyword? theme-or-opacity)
-        (resolve-color color theme-or-opacity nil)
-        (resolve-color color nil theme-or-opacity)))
-     ([color theme opacity]
-      (let [suffix         (cond opacity         50
-                                 (= :dark theme) 60
-                                 :else           50)
-            hex?           (not (keyword? color))
-            resolved-color (cond hex?                                 color
-                                 (hex-string? (get colors-map color)) (get colors-map color)
-                                 :else                                (get-in colors-map
-                                                                              [color suffix]))]
-        (if opacity
-          (alpha resolved-color (/ opacity 100))
-          resolved-color))))))
+  (memoize do-resolve-color))
 
 (def ^{:deprecated true :superseded-by "resolve-colors"}
      custom-color
```

Reply:
Yeah we can normalise the network colors, if kind of breaks the mould a bit however as there is no -60 network colors in the foundations file.

Another thing is we need the hex string for community colors as that's how they come from the backend

Reply:
> Yeah we can normalise the network colors, if kind of breaks the mould a bit however as there is no -60 network colors in the foundations file.

Exactly, tradeoffs.

> Another thing is we need the hex string for community colors as that's how they come from the backend

Could you explain this part @J-Son89? I kind of understand.

----

About the implementation I suggested, another improvement is to avoid calling twice `(get colors-map color)`. This one also passes the unit tests.

```clojure
(defn- do-resolve-color
  ([color theme]
   (do-resolve-color color theme nil))
  ([color theme opacity]
   (let [suffix   (cond opacity         50
                        (= :dark theme) 60
                        :else           50)
         resolved (if (hex-string? color)
                    color
                    (let [maybe-resolved (get colors-map color)]
                      (if (hex-string? maybe-resolved)
                        maybe-resolved
                        (get-in colors-map [color suffix]))))]
     (if opacity
       (alpha resolved (/ opacity 100))
       resolved))))

(def resolve-color (memoize do-resolve-color))
```
```

Reply:
Oh, apologies I also meant to revert this - 
"""
(if (keyword? theme-or-opacity)
-        (resolve-color color theme-or-opacity nil)
-        (resolve-color color nil theme-
"""
It was just an idea I was experimenting with before discussing with you. D'oh! 🤦‍♂️

Reply:
With community colors, the creator of a community can make that any hex string they like. However with user colors, network colors and chat colors these are all from a predefined palette decides by the design team. Hopefully that's a bit clearer 😅

Reply:
It is, thank you @J-Son89!